File: frontend/FleetForge/src/app/driver/service/driver-dashboard/driver-dashboard.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, forkJoin } from 'rxjs';

export interface UpcomingRideDTO {
  rideId: number;
  passengerName: string;
  passengerImage: string;
  startAddress: string;
  endAddress: string;
  startTime: string;
  cost: number;
}

export interface DailyStatsDTO {
  totalEarnings: number;
  hoursWorked: number;
  ridesCompleted: number;
}

export interface DashboardData {
  upcomingRides: UpcomingRideDTO[];
  dailyStats: DailyStatsDTO;
}

@Injectable({ providedIn: 'root' })
export class DriverDashboardService {
  private apiUrl = 'http://localhost:8080/api/drivers';

  constructor(private http: HttpClient) {}

  getUpcomingRides(): Observable<UpcomingRideDTO[]> {
    return this.http.get<UpcomingRideDTO[]>(`${this.apiUrl}/upcoming-rides`);
  }

  getDailyStats(): Observable<DailyStatsDTO> {
    return this.http.get<DailyStatsDTO>(`${this.apiUrl}/daily-stats`);
  }

  getDashboardData(): Observable<DashboardData> {
    return forkJoin({
      upcomingRides: this.getUpcomingRides(),
      dailyStats: this.getDailyStats()
    });
  }
}

File: backend/FleetForge/src/main/java/com/team18/FleetForge/service/DriverDashboardService.java
package com.team18.FleetForge.service;

import com.team18.FleetForge.dto.ride.lifecycle.UpcomingRideDTO;
import com.team18.FleetForge.model.enums.RideStatus;
import com.team18.FleetForge.model.ride.Ride;
import com.team18.FleetForge.model.users.Driver;
import com.team18.FleetForge.repository.DriverRepository;
import com.team18.FleetForge.repository.RideRepository;
import com.team18.FleetForge.dto.driver.DailyStatsDTO;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class DriverDashboardService {

    private final RideRepository rideRepository;
    private final DriverRepository driverRepository;

    public List<UpcomingRideDTO> getUpcomingRides(Long id) {

        List<Ride> upcomingRides = rideRepository.findActiveRidesByDriverId(id);

        return upcomingRides.stream()
                .map(this::buildUpcomingRideDTO)
                .collect(Collectors.toList());

    }

    public DailyStatsDTO getDailyStats(Long driverId) {
        Driver driver = driverRepository.findById(driverId)
                .orElseThrow(() -> new RuntimeException("Driver not found"));

        LocalDateTime startOfDay = LocalDateTime.now().toLocalDate().atStartOfDay();
        LocalDateTime endOfDay = startOfDay.plusDays(1);

        List<Ride> completedRidesToday = rideRepository.findAllByDriverAndStatusAndEndTimeBetween(
                driver,
                RideStatus.COMPLETED,
                startOfDay,
                endOfDay
        );

        Double totalEarnings = completedRidesToday.stream()
                .mapToDouble(Ride::getTotalCost)
                .sum();

        Double hoursWorked = 5.0; 

        Integer ridesCompleted = completedRidesToday.size();

        return DailyStatsDTO.builder()
                .totalEarnings(totalEarnings)
                .hoursWorked(hoursWorked)
                .ridesCompleted(ridesCompleted)
                .build();
    }


    private UpcomingRideDTO buildUpcomingRideDTO(Ride ride) {
        return UpcomingRideDTO.builder()
                .rideId(ride.getId())
                .passengerName(ride.getPassenger().getFirstName() + " " + ride.getPassenger().getLastName())
                .passengerImage(ride.getPassenger().getProfilePicture())
                .startAddress(ride.getStartAddress())
                .endAddress(ride.getEndAddress())
                .startTime(ride.getStartTime())
                .cost(ride.getTotalCost())
                .build();
    }
}
